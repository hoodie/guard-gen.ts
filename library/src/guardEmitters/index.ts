export * from './interface-emitter'
export * from './type-emitter'