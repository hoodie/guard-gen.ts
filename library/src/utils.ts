import { Console } from "console";

export const comment = (...x: any[]) => logger.log.apply(console, [`//`].concat(x) as any);

const utilLogger = new Console({ stdout: process.stderr, stderr: process.stderr });
const dumpLogger = {
    log(...args: any) { },
    info(...args: any) { },
    dir(...args: any) { },
    table(...args: any) { },
    debug(...args: any) { },
    trace(...args: any) { },
    warn(...args: any) { },
    error(...args: any) { },
};

const loggerSettings = { on: false }
export let logger = dumpLogger;
export const toggleLogger = (on: boolean) => {
    loggerSettings.on = on;
    logger = on ? utilLogger : dumpLogger;
}

export const dir = (x: any) => logger.dir(x, { colors: true, depth: 4 });

export const odir = (x: any) =>
    loggerSettings.on &&
    dir(
        JSON.parse(
            JSON.stringify(x, (k, v) => {
                if (k !== 'parent') {
                    return v;
                }
            })
        )
    );

export const capitalize = (s: string): string => s[0].toUpperCase() + s.slice(1);
