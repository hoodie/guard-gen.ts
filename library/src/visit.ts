import ts, { isTypeNode, SyntaxKind, visitNodes } from 'typescript';
import { logger, odir } from './utils';

export interface NodeInfo {
    name: string;
    valueCheck: string;
    typeName: string;
    typeArguments?: ts.NodeArray<ts.TypeNode>;
    isOptional: boolean;
}

interface VisitorContext<T extends ts.TypeNode = ts.TypeNode> extends NodeInfo {
    node: ts.HasType;
    exportedSymbols: string[];
}

export const isExported = (node: ts.Statement) => node.modifiers?.some(({ kind }) => kind === SyntaxKind.ExportKeyword);

interface HasType {
    type: ts.TypeNode;
}

const hasType = (node: any): node is HasType => node.type && ts.isTypeNode(node.type);

function assertHasType(node: any): asserts node is HasType {
    if (!hasType(node)) {
        throw Error('node does not have a type');
    }
}

type Visitor<T extends ts.TypeNode = ts.TypeNode> = (ctx: VisitorContext<T>) => NodeInfo;

const typeCheck = (typ: ts.TypeNode): string => {
    const { valueCheck, typeName } = visitNode({ node: { type: typ } as any });
    return valueCheck ? `(x) => ${valueCheck}` : `/* unimplemented for ${typeName} */`;
};

export const tryName = (node: ts.Node & any) => node.name.escapedText || node.name;

const debugNode = (node: ts.Node & HasType) => {
    const info = {
        type: ts.SyntaxKind[node.type.kind],
        name: tryName(node),
        kind: ts.SyntaxKind[node.kind],
    };
    logger.log(`ðŸ”‘ ${info.kind} '${info.name}' = ${info.type}`);
    // odir(node)
};

// string, number, object
const toNodeInfo: Visitor = ({ name, typeName, typeArguments, valueCheck, isOptional }) => ({
    name,
    typeName,
    typeArguments,
    valueCheck,
    isOptional,
});

// string, number, object
const visitPrimitive: Visitor = ({ name, typeName, typeArguments, isOptional }) => ({
    name,
    // name: (logger.debug('    visitPrimitive', typeName), name),
    typeName,
    typeArguments,
    valueCheck: `typeof ${name} === '${typeName}'`,
    isOptional,
});

// rootVisitor
export function visitNode({ node, name, exportedSymbols }: Partial<VisitorContext>): NodeInfo {
    logger.debug('  visitNode');
    assertHasType(node);

    const ctx = {
        name: name || 'x',
        node,
        typeArguments: ((node as any) as ts.NodeWithTypeArguments).typeArguments,
        valueCheck: `true /* unimplemented for ${SyntaxKind[node.type.kind]} "${name}" */`,
        isOptional: !!(node as any).questionToken,
        exportedSymbols: exportedSymbols || [],
    };

    debugNode(node);

    if (node.type.kind === SyntaxKind.NumberKeyword) {
        return visitPrimitive({ ...ctx, typeName: 'number' });
    } else if (node.type.kind === SyntaxKind.StringKeyword) {
        return visitPrimitive({ ...ctx, typeName: 'string' });
    } else if (node.type.kind === SyntaxKind.ObjectKeyword) {
        return visitPrimitive({ ...ctx, typeName: 'object' });
    } else if (node.type.kind === SyntaxKind.AnyKeyword) {
        return toNodeInfo({ ...ctx, typeName: 'any', valueCheck: `true` });
    } else if (ts.isLiteralTypeNode(node.type) && ts.isLiteralExpression(node.type.literal)) {
        return toNodeInfo({
            ...ctx,
            typeName: `'${node.type.literal.text}'`,
            valueCheck: `${name} === '${node.type.literal.text}'`,
        });
    } else if (ts.isArrayTypeNode(node.type)) {
        const elementTypeName = visitNode({
            node: { type: node.type.elementType } as any,
        }).typeName;
        const elementCheck = typeCheck(node.type.elementType);

        return toNodeInfo({
            ...ctx,
            typeName: `Array<${elementTypeName}>`,
            valueCheck: `(Array.isArray(${ctx.name}) && ${ctx.name}.every(${elementCheck}))`,
        });
    } else if (ts.isMappedTypeNode(node.type)) {
        logger.debug('ðŸ¤¯');
        odir(node);
        throw new Error('unhandled case: typereference without identifier');
    } else if (ts.isTypeReferenceNode(node.type)) {
        if (ts.isIdentifier(node.type.typeName)) {
            const typeName = node.type.typeName.escapedText as string;
            const exported = exportedSymbols?.includes(typeName);
            const preCheck = ctx.isOptional ? `${name} && ` : ''; // because a || doesn't cut it
            const valueCheck = exported ? `${preCheck}is${typeName}(${name})` : `true /* ${name}: ${typeName}*/`;

            logger.debug('ðŸ¤”', {
                isExpressionWithTypeArguments: ts.isExpressionWithTypeArguments(node),
                info: toNodeInfo({
                    ...ctx,
                    typeName,
                    typeArguments: node.type.typeArguments,
                    valueCheck,
                }),
            });
            odir(node);
            return toNodeInfo({
                ...ctx,
                typeName,
                typeArguments: node.type.typeArguments,
                valueCheck,
            });
        } else {
            throw new Error('unhandled case: typereference without identifier');
        }
    } else if (ts.isUnionTypeNode(node.type)) {
        const l = node.type.types.map((t) => visitNode({ ...ctx, node: { type: t } as any }));
        const check = l.map((t) => t.valueCheck).join(' || ');
        return toNodeInfo({
            ...ctx,
            typeName: l.map((t) => t.typeName).join(' | '),
            valueCheck: `(${check})`,
        });
    } else if (ts.isParenthesizedTypeNode(node.type)) {
        const newNode = { ...node, type: node.type.type };
        return visitNode({ ...ctx, node: newNode });
        // } else if (ts.isPropertySignature(node) && ts.isFunctionTypeNode(node.type) && ts.isIdentifier(node.name)) {
        //     const { valueCheck } = visitPrimitive({ ...ctx, typeName: 'function' });
        //     return toNodeInfo({
        //         ...ctx,
        //         typeName: `unhandled typeName(${SyntaxKind[node.type.kind]})`,
        //         valueCheck: `${valueCheck} && (name.length >= ${node.type.parameters.length})`,
        //     });
    } else {
        return toNodeInfo({
            ...ctx,
            typeName: `unhandled typeName(${SyntaxKind[node.type.kind]})`,
            valueCheck: ctx.valueCheck,
        });
    }
}

///////////// new
/*

type Visitor2<T extends ts.TypeNode = ts.TypeNode> = (
    node: { type: T },
    ctx: VisitorContext<T>
) => NodeInfo | undefined;

const primitiveTypeNames = {
    [SyntaxKind.NumberKeyword]: `number`,
    [SyntaxKind.StringKeyword]: 'string',
    [SyntaxKind.ObjectKeyword]: 'object',
};

export interface PrimitiveNode extends ts.Node {
    readonly type: PrimitiveTypeNode;
}

export interface PrimitiveTypeNode extends ts.TypeNode {
    readonly kind: SyntaxKind.NumberKeyword | SyntaxKind.StringKeyword | SyntaxKind.ObjectKeyword;
}

export const isPrimitiveType = (node: ts.TypeNode): node is PrimitiveTypeNode => {
    return (
        node.kind === SyntaxKind.NumberKeyword ||
        node.kind === SyntaxKind.StringKeyword ||
        node.kind === SyntaxKind.ObjectKeyword
    );
};

const isPrimitiveNode = (node: { type: ts.TypeNode }): node is PrimitiveNode => {
    return isPrimitiveType(node.type);
};

type NodeOfType<T extends ts.SyntaxKind> = { type: { kind: T } & ts.TypeNode };
type x = ts.HasTypedkts.HasType['kind']>

const isSomeNode = <K extends ts.SyntaxKind>(node: NodeOfType<any>, kind: K): node is NodeOfType<K> => {
    return node.type.kind === kind;
};

const visitorPrimitive: Visitor2<PrimitiveTypeNode> = (node, ctx): NodeInfo => {
    const typeName = primitiveTypeNames[node.type.kind];
    return {
        ...ctx,
        typeName,
        valueCheck: `typeof ${ctx.name} === '${typeName}'`,
    };
};

const visitorAny: Visitor2 = (node, ctx): NodeInfo | undefined => {
    if (node.type.kind === SyntaxKind.AnyKeyword) {
        return toNodeInfo({ ...ctx, typeName: 'any', valueCheck: `true` });
    }
};

const visitorLiteral: Visitor2<ts.LiteralTypeNode> = (node, ctx): NodeInfo | undefined => {
    if (ts.isLiteralTypeNode(node.type) && ts.isLiteralExpression(node.type.literal)) {
        return toNodeInfo({
            ...ctx,
            typeName: `'${node.type.literal.text}'`,
            valueCheck: `${ctx.name} === '${node.type.literal.text}'`,
        });
    }
};

export function visitNode2({ node, name, exportedSymbols }: Partial<VisitorContext>): NodeInfo | undefined {
    logger.debug('  visitNode');
    if (!hasType(node)) {
        throw new Error('only nodes with type allowed');
    }

    const ctx = {
        name: name || 'x',
        node,
        typeName: 'unknown',
        typeArguments: ((node as any) as ts.NodeWithTypeArguments).typeArguments,
        valueCheck: `true /* unimplemented for ${SyntaxKind[node.type.kind]} "${name}" */`,
        isOptional: !!(node as any).questionToken,
        exportedSymbols: exportedSymbols || [],
    };

    if (isPrimitiveNode(node)) {
        return visitorPrimitive(node, ctx);
    //} else if ( ts.isLiteralTypeNode(node.type)) {
    } else if (
        isSomeNode(node, SyntaxKind.LiteralType) || 
        isSomeNode(node, SyntaxKind.SyntaxList)
        ) {
        return visitorLiteral(node, ctx);
    } else {
        return visitorAny(node, ctx);
    }
}
*/