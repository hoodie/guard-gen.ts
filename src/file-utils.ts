import { existsSync, statSync } from 'fs';
import { basename, dirname, join, relative } from 'path';
import { comment } from './utils';

const dtsBasename = (inputFile: string): string =>
    basename(basename(inputFile, '.d.ts'), '.ts');

const makePathImportable = (tsPath: string): string => {
    const importFilePath = join(dirname(tsPath), dtsBasename(tsPath));
    if (importFilePath.startsWith('../')) {
        return importFilePath;
    } else {
        return `./${importFilePath}`;
    }
};

const guardFileName = (inputFile: string): string =>
    dtsBasename(inputFile) + '.guards.ts';



// where does the outfile find the inputFile
export const importPath = (inputFile: string, outfile: string): string =>
    makePathImportable(relative(dirname(outfile), inputFile));

// if no outfile path is given
export const derivedOutfilePath = (inputFile: string): string =>
    join(dirname(inputFile), guardFileName(inputFile));

// if outfile path is either a dir or a file
export const outfilePath = (inputFile: string, targetPath: string): string => {
    if (existsSync(targetPath) && statSync(targetPath).isDirectory()) {
        return join(targetPath, guardFileName(inputFile));
    } else {
        return targetPath;
    }
};
