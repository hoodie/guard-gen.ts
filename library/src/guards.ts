/* eslint-disable @typescript-eslint/no-use-before-define */
import ts from 'typescript';
import { isExported, tryName } from './visit';
import { logger } from './utils';

import { emitInterfaceGuard, emitTypeGuard } from "./guardEmitters";

const emitGuard = (node: ts.Node, exportedSymbols: string[], config: EmitterConfig): string => {
    logger.debug(`generateGuard "${tryName(node)}"`);

    if (ts.isInterfaceDeclaration(node)) {
        return emitInterfaceGuard(node, exportedSymbols, config);

    } else if (ts.isTypeAliasDeclaration(node)) {
        return emitTypeGuard(node, exportedSymbols);

    } else {
        return `// `;
    }
};

const publicStatements = (sourceFile: ts.SourceFile): { statements: ts.Statement[]; names: string[] } => {
    const names: string[] = [];
    const statements: ts.Statement[] = [];

    sourceFile.statements.filter(isExported).forEach((statement) => {
        if (ts.isInterfaceDeclaration(statement) || ts.isTypeAliasDeclaration(statement)) {
            statements.push(statement);
            names.push(statement.name.escapedText as string);
        }
    });

    return { statements, names };
};

// public

export interface EmitterConfig {
    // path to import types from
    importFrom?: string;

    // generate warning code for every single property
    embedWarnings: boolean;
}

export function emitImportLine(sourceFile: ts.SourceFile, importFrom: string): string {
    logger.debug('generateImportLine', { importFrom });
    const { names } = publicStatements(sourceFile);
    return `import {${names.sort().join(', ')}} from '${importFrom}';`;
}

export function emitGuards(sourceFile: ts.SourceFile, config: EmitterConfig): string[] {
    logger.debug('generateGuards', { config });
    const guards: string[] = [];
    const { statements, names: exportedSymbols } = publicStatements(sourceFile);
    statements.forEach((node) => {
        const guard = emitGuard(node, exportedSymbols, config);
        guards.push(guard);
    });
    return guards;
}
