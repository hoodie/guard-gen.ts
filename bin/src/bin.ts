import programm from 'commander';
import * as fs from 'fs';
import ts from 'typescript';
import { importPath, outfilePath } from './file-utils';

const openFile = (path: string): string => fs.readFileSync(path, 'utf8');
const error = (message: string) => console.error(`ERROR: ${message}`);

    /*
programm
    .version('0.1.0')
    .name('guardner')
    .command('generate [FILE]')
    .alias('gen')
    .description('generate guards from given ts file')
    .option('-w, --warners', 'embed code that produces warnings')
    .option('-g, --guardsfile', 'put a .guards.ts file next to your input')
    .option('-o, --outfile [FILE]', 'path to file to generate')
    .action((inputFile, { warners, outfile }) => {
        const embedWarnings = !!warners;
        if (!!inputFile) {
            const outputFile = outfilePath(inputFile, outfile);

            let generated: Generated;

            if (outfile) {
                const importFrom = importPath(inputFile, outputFile);
                generated = generateFrom(inputFile, {
                  importFrom,
                  embedWarnings
                });
                writeGuardsFile(generated, outfile);
            } else {
                generated = generateFrom(inputFile, { embedWarnings });
                printGuards(generated);
            }

        } else {
            error('please pass a typescript file to generate guards for');
        }
    });


if (!process.argv.slice(2).length) {
    programm.outputHelp();

} else {
    programm.parse(process.argv);

}

*/