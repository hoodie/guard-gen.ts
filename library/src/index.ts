import * as fs from 'fs';
import ts from 'typescript';
import { generateGuards, generateImportLine, GeneratorConfig } from './guards';

const openFile = (path: string): string => fs.readFileSync(path, 'utf8');
const error = (message: string) => console.error(`ERROR: ${message}`);

export interface Generated {
    imports?: string;
    guards: string[];
}

export function generateFrom(sourceFilePath: string, config: GeneratorConfig): Generated {
    const sourceText = openFile(sourceFilePath);
    const sourceFile = ts.createSourceFile(
        'x.ts',
        sourceText,
        ts.ScriptTarget.Latest,
        true
    );

    const { importFrom } = config;
    const imports = importFrom && generateImportLine(sourceFile, importFrom);
    const guards = generateGuards(sourceFile, config);

    return { imports, guards };
}

const printGuards = (generated: Generated) =>
    console.log(concatGuards(generated));

const concatGuards = ({ imports, guards }: Generated): string =>
    [imports].concat(guards).join('\n');

const writeGuardsFile = (generated: Generated, path: string) =>
    fs.writeFileSync(path, concatGuards(generated), { encoding: 'utf8' });

    /*
programm
    .version('0.1.0')
    .name('guardner')
    .command('generate [FILE]')
    .alias('gen')
    .description('generate guards from given ts file')
    .option('-w, --warners', 'embed code that produces warnings')
    .option('-g, --guardsfile', 'put a .guards.ts file next to your input')
    .option('-o, --outfile [FILE]', 'path to file to generate')
    .action((inputFile, { warners, outfile }) => {
        const embedWarnings = !!warners;
        if (!!inputFile) {
            const outputFile = outfilePath(inputFile, outfile);

            let generated: Generated;

            if (outfile) {
                const importFrom = importPath(inputFile, outputFile);
                generated = generateFrom(inputFile, {
                  importFrom,
                  embedWarnings
                });
                writeGuardsFile(generated, outfile);
            } else {
                generated = generateFrom(inputFile, { embedWarnings });
                printGuards(generated);
            }

        } else {
            error('please pass a typescript file to generate guards for');
        }
    });


if (!process.argv.slice(2).length) {
    programm.outputHelp();

} else {
    programm.parse(process.argv);

}

*/