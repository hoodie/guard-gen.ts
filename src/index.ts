#!/bin/env node
import * as fs from 'fs';
import ts from 'typescript';
import { generateGuards, generateImportLine } from './guards';

const openFile = (path: string): string => fs.readFileSync(path, 'utf8');

// const sourceFilePath = './test/example-types.ts';
const sourceFilePath = process.argv[2];
const sourceText = openFile(sourceFilePath);
const sourceFile = ts.createSourceFile(
  "x.ts",
  sourceText,
  ts.ScriptTarget.Latest,
  true
);

const imports = generateImportLine(sourceFile, sourceFilePath);
const guards = generateGuards(sourceFile);

console.log(imports);

console.log(guards.join('\n'));