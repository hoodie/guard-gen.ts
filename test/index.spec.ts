import "jest"
import { generateFrom } from "../src";

describe('generator', () => {
    it('generates the same code as yesterday without warnings', () => {
        const generated = generateFrom('./test/fixtures/example-types.ts', {embedWarnings: false});
        expect(JSON.stringify(generated)).toMatchSnapshot()

    })
    it('generates the same code as yesterday with warnings', () => {
        const generated = generateFrom('./test/fixtures/example-types.ts', {embedWarnings: true});
        expect(JSON.stringify(generated)).toMatchSnapshot()

    })
});