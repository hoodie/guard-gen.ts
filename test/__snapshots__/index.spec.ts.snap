// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generator generates the same code as yesterday with warnings 1`] = `"{\\"guards\\":[\\"\\\\n// generated typeguard for Foo\\\\nexport const isFoo = (maybeFoo: any): maybeFoo is Foo => {\\\\n    const {name, value, amount, more, muchMore} = maybeFoo;\\\\n\\\\n    return ((name) => {\\\\n        const nameChecksOut = name === 'foo' /* name: 'foo'*/;\\\\n        if(!nameChecksOut) {console.warn(\\\\\\"name is not a propper 'foo'\\\\\\")}\\\\n        return nameChecksOut;\\\\n        })(name) &&\\\\n        ((value) => {\\\\n        const valueChecksOut = typeof value === 'string' /* value: string*/;\\\\n        if(!valueChecksOut) {console.warn(\\\\\\"value is not a propper string\\\\\\")}\\\\n        return valueChecksOut;\\\\n        })(value) &&\\\\n        ((amount) => {\\\\n        const amountChecksOut = !(amount) || typeof amount === 'number' /* amount: number?*/;\\\\n        if(!amountChecksOut) {console.warn(\\\\\\"amount is not a propper number\\\\\\")}\\\\n        return amountChecksOut;\\\\n        })(amount) &&\\\\n        ((more) => {\\\\n        const moreChecksOut = typeof more === 'object' /* more: object*/;\\\\n        if(!moreChecksOut) {console.warn(\\\\\\"more is not a propper object\\\\\\")}\\\\n        return moreChecksOut;\\\\n        })(more) &&\\\\n        ((muchMore) => {\\\\n        const muchMoreChecksOut = true /* muchMore: any*/;\\\\n        if(!muchMoreChecksOut) {console.warn(\\\\\\"muchMore is not a propper any\\\\\\")}\\\\n        return muchMoreChecksOut;\\\\n        })(muchMore);\\\\n};\\\\n\\\\n\\",\\"\\\\n// generated typeguard for Bar\\\\nexport const isBar = (maybeBar: any): maybeBar is Bar => {\\\\n    const {name, list, maybesomething} = maybeBar;\\\\n\\\\n    return ((name) => {\\\\n        const nameChecksOut = name === 'bar' /* name: 'bar'*/;\\\\n        if(!nameChecksOut) {console.warn(\\\\\\"name is not a propper 'bar'\\\\\\")}\\\\n        return nameChecksOut;\\\\n        })(name) &&\\\\n        ((list) => {\\\\n        const listChecksOut = (Array.isArray(list) && list.every((x) => typeof x === 'number')) /* list: Array<number>*/;\\\\n        if(!listChecksOut) {console.warn(\\\\\\"list is not a propper Array<number>\\\\\\")}\\\\n        return listChecksOut;\\\\n        })(list) &&\\\\n        ((maybesomething) => {\\\\n        const maybesomethingChecksOut = (maybesomething === 'something' || maybesomething === 'nothing') /* maybesomething: 'something' | 'nothing'*/;\\\\n        if(!maybesomethingChecksOut) {console.warn(\\\\\\"maybesomething is not a propper 'something' | 'nothing'\\\\\\")}\\\\n        return maybesomethingChecksOut;\\\\n        })(maybesomething);\\\\n};\\\\n\\\\n\\",\\"\\\\n// generated typeguard for Direction\\\\nexport const isDirection = (maybeDirection: any): maybeDirection is Direction =>\\\\n    (maybeDirection === 'up' || maybeDirection === 'down');\\",\\"\\\\n// generated typeguard for FooOrBar\\\\nexport const isFooOrBar = (maybeFooOrBar: any): maybeFooOrBar is FooOrBar =>\\\\n    (isFoo(maybeFooOrBar) || isBar(maybeFooOrBar));\\",\\"\\\\n// generated typeguard for Foobar\\\\nexport const isFoobar = (maybeFoobar: any): maybeFoobar is Foobar => {\\\\n    const {foo, bar, extras, names, direction, directions, numbers, parts} = maybeFoobar;\\\\n\\\\n    return ((foo) => {\\\\n        const fooChecksOut = isFoo(foo) /* foo: Foo*/;\\\\n        if(!fooChecksOut) {console.warn(\\\\\\"foo is not a propper Foo\\\\\\")}\\\\n        return fooChecksOut;\\\\n        })(foo) &&\\\\n        ((bar) => {\\\\n        const barChecksOut = isBar(bar) /* bar: Bar*/;\\\\n        if(!barChecksOut) {console.warn(\\\\\\"bar is not a propper Bar\\\\\\")}\\\\n        return barChecksOut;\\\\n        })(bar) &&\\\\n        ((extras) => {\\\\n        const extrasChecksOut = !(extras) || true /* unimplemented for TypeLiteral \\\\\\"extras\\\\\\" */ /* extras: unhandled typeName(TypeLiteral)?*/;\\\\n        if(!extrasChecksOut) {console.warn(\\\\\\"extras is not a propper unhandled typeName(TypeLiteral)\\\\\\")}\\\\n        return extrasChecksOut;\\\\n        })(extras) &&\\\\n        ((names) => {\\\\n        const namesChecksOut = !(names) || (Array.isArray(names) && names.every((x) => typeof x === 'string')) /* names: Array<string>?*/;\\\\n        if(!namesChecksOut) {console.warn(\\\\\\"names is not a propper Array<string>\\\\\\")}\\\\n        return namesChecksOut;\\\\n        })(names) &&\\\\n        ((direction) => {\\\\n        const directionChecksOut = isDirection(direction) /* direction: Direction*/;\\\\n        if(!directionChecksOut) {console.warn(\\\\\\"direction is not a propper Direction\\\\\\")}\\\\n        return directionChecksOut;\\\\n        })(direction) &&\\\\n        ((directions) => {\\\\n        const directionsChecksOut = (Array.isArray(directions) && directions.every((x) => true /* undefined: Direction*/)) /* directions: Array<Direction>*/;\\\\n        if(!directionsChecksOut) {console.warn(\\\\\\"directions is not a propper Array<Direction>\\\\\\")}\\\\n        return directionsChecksOut;\\\\n        })(directions) &&\\\\n        ((numbers) => {\\\\n        const numbersChecksOut = !(numbers) || (Array.isArray(numbers) && numbers.every((x) => typeof x === 'number')) /* numbers: Array<number>?*/;\\\\n        if(!numbersChecksOut) {console.warn(\\\\\\"numbers is not a propper Array<number>\\\\\\")}\\\\n        return numbersChecksOut;\\\\n        })(numbers) &&\\\\n        ((parts) => {\\\\n        const partsChecksOut = true /* parts: PrivateInterface*/ /* parts: PrivateInterface*/;\\\\n        if(!partsChecksOut) {console.warn(\\\\\\"parts is not a propper PrivateInterface\\\\\\")}\\\\n        return partsChecksOut;\\\\n        })(parts);\\\\n};\\\\n\\\\n\\"]}"`;

exports[`generator generates the same code as yesterday without warnings 1`] = `"{\\"guards\\":[\\"\\\\n// generated typeguard for Foo\\\\nexport const isFoo = (maybeFoo: any): maybeFoo is Foo => {\\\\n    const {name, value, amount, more, muchMore} = maybeFoo;\\\\n\\\\n    return name === 'foo' /* name: 'foo'*/ &&\\\\n        typeof value === 'string' /* value: string*/ &&\\\\n        !(amount) || typeof amount === 'number' /* amount: number?*/ &&\\\\n        typeof more === 'object' /* more: object*/ &&\\\\n        true /* muchMore: any*/;\\\\n};\\\\n\\\\n\\",\\"\\\\n// generated typeguard for Bar\\\\nexport const isBar = (maybeBar: any): maybeBar is Bar => {\\\\n    const {name, list, maybesomething} = maybeBar;\\\\n\\\\n    return name === 'bar' /* name: 'bar'*/ &&\\\\n        (Array.isArray(list) && list.every((x) => typeof x === 'number')) /* list: Array<number>*/ &&\\\\n        (maybesomething === 'something' || maybesomething === 'nothing') /* maybesomething: 'something' | 'nothing'*/;\\\\n};\\\\n\\\\n\\",\\"\\\\n// generated typeguard for Direction\\\\nexport const isDirection = (maybeDirection: any): maybeDirection is Direction =>\\\\n    (maybeDirection === 'up' || maybeDirection === 'down');\\",\\"\\\\n// generated typeguard for FooOrBar\\\\nexport const isFooOrBar = (maybeFooOrBar: any): maybeFooOrBar is FooOrBar =>\\\\n    (isFoo(maybeFooOrBar) || isBar(maybeFooOrBar));\\",\\"\\\\n// generated typeguard for Foobar\\\\nexport const isFoobar = (maybeFoobar: any): maybeFoobar is Foobar => {\\\\n    const {foo, bar, extras, names, direction, directions, numbers, parts} = maybeFoobar;\\\\n\\\\n    return isFoo(foo) /* foo: Foo*/ &&\\\\n        isBar(bar) /* bar: Bar*/ &&\\\\n        !(extras) || true /* unimplemented for TypeLiteral \\\\\\"extras\\\\\\" */ /* extras: unhandled typeName(TypeLiteral)?*/ &&\\\\n        !(names) || (Array.isArray(names) && names.every((x) => typeof x === 'string')) /* names: Array<string>?*/ &&\\\\n        isDirection(direction) /* direction: Direction*/ &&\\\\n        (Array.isArray(directions) && directions.every((x) => true /* undefined: Direction*/)) /* directions: Array<Direction>*/ &&\\\\n        !(numbers) || (Array.isArray(numbers) && numbers.every((x) => typeof x === 'number')) /* numbers: Array<number>?*/ &&\\\\n        true /* parts: PrivateInterface*/ /* parts: PrivateInterface*/;\\\\n};\\\\n\\\\n\\"]}"`;
